src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -646,9 +674,9 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
             }
         }
 
-        private static BoundExpression BindDefaultLiteral(ExpressionSyntax node)
+        private static BoundDefaultExpression BindDefaultLiteral(ExpressionSyntax node)
         {
-            return new BoundDefaultExpression(node, type: null);
+            return new BoundDefaultExpression(node, targetType: null, constantValueOpt: null, type: null);
```

Comment:
>BoundDefaultExpression [](start = 23, length = 22)

I think we should have a special node BoundDefaultLiteral and leave BoundDefaultExpression exclusively for ```default(T)```

Reply:
`default` currently gets bound to `BoundDefaultExpression` (then a `BoundConvertedDefaultExpression` after conversion) and `default(T)` directly binds to `BoundConvertedDefaultExpression`. 
Are you suggesting to introduce a third kind of bound node, or merely renaming the nodes `BoundDefaultExpression` -> `BoundDefaultLiteral` and `BoundConvertedDefaultExpression` -> `BoundDefaultExpression`?

---
In reply to: [314080570](https://github.com/dotnet/roslyn/pull/37596#discussion_r314080570) [](ancestors = 314080570)

Reply:
I guess simply introducing BoundDefaultLiteral node that gets transformed to BoundDefaultExpression might work, unless we need more information than BoundDefaultExpression has right now. Then we would need a third node BoundConvertedDefaultLiteral

---
In reply to: [314087380](https://github.com/dotnet/roslyn/pull/37596#discussion_r314087380) [](ancestors = 314087380,314080570)

Reply:
Aside from naming, the only difference I'm introducing is that the converted node stores a natural type, which the current `BoundDefaultExpression` doesn't. I don't think it's useful to introduce a third node for that purpose.

---
In reply to: [314090290](https://github.com/dotnet/roslyn/pull/37596#discussion_r314090290) [](ancestors = 314090290,314087380,314080570)

Reply:
> Aside from naming, the only difference I'm introducing is that the converted node stores a natural type, which the current BoundDefaultExpression doesn't. I don't think it's useful to introduce a third node for that purpose.

I find it confusing to use Converted node for default(T) and have meaningless NaturalType property for it. Also, it is advised to create only leave nodes not the base ones. We weren't always following this, but the motivation is clarity and code hygiene. If you follow that, two nodes just wouldn't work anyway. 

---
In reply to: [314095338](https://github.com/dotnet/roslyn/pull/37596#discussion_r314095338) [](ancestors = 314095338,314090290,314087380,314080570)

Reply:
I think, Default literal and corresponding converted node shouldn't have TargetType property, default expression node shouldn't have NaturalType property, Defaultliteral node should its Type property always be null, not even settable through constructor. Then I can easily make sense what does the tree mean and how it came to be.

---
In reply to: [314099454](https://github.com/dotnet/roslyn/pull/37596#discussion_r314099454) [](ancestors = 314099454,314095338,314090290,314087380,314080570)